plugins/inputs/intel_rdt/intel_rdt.go

```diff
@@ -0,0 +1,552 @@
+// +build !windows
+
+package intel_rdt
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	timestampFormat           = "2006-01-02 15:04:05"
+	defaultSamplingInterval   = 10
+	pqosInitOutputLinesNumber = 4
+	numberOfMetrics           = 6
+	secondsDenominator        = 10
+)
+
+var pqosMetricOrder = map[int]string{
+	0: "IPC",        // Instructions Per Cycle
+	1: "LLC_Misses", // Cache Misses
+	2: "LLC",        // L3 Cache Occupancy
+	3: "MBL",        // Memory Bandwidth on Local NUMA Node
+	4: "MBR",        // Memory Bandwidth on Remote NUMA Node
+	5: "MBT",        // Total Memory Bandwidth
+}
+
+type IntelRDT struct {
+	PqosPath         string   `toml:"pqos_path"`
+	Cores            []string `toml:"cores"`
+	Processes        []string `toml:"processes"`
+	SamplingInterval int32    `toml:"sampling_interval"`
+	ShortenedMetrics bool     `toml:"shortened_metrics"`
+
+	Log              telegraf.Logger  `toml:"-"`
+	Publisher        Publisher        `toml:"-"`
+	Processor        ProcessesHandler `toml:"-"`
+	stopPQOSChan     chan bool
+	quitChan         chan struct{}
+	errorChan        chan error
+	parsedCores      []string
+	processesPIDsMap map[string]string
+	cancel           context.CancelFunc
+	wg               sync.WaitGroup
+}
+
+type processMeasurement struct {
+	name        string
+	measurement string
+}
+
+// All gathering is done in the Start function
+func (r *IntelRDT) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (r *IntelRDT) Description() string {
+	return "Intel Resource Director Technology plugin"
+}
+
+func (r *IntelRDT) SampleConfig() string {
+	return `
+	## Optionally set sampling interval to Nx100ms. 
+	## This value is propagated to pqos tool. Interval format is defined by pqos itself.
+	## If not provided or provided 0, will be set to 10 = 10x100ms = 1s.
+	# sampling_interval = "10"
```

Comment:
I find this a little unclear. is this the number of samples per second?

Reply:
FYI there is also a config.Duration type

Reply:
Thank you for the info about config.Duration. We decided to use sampling_interval to be consistent with pqos tool itself. We provide this configuration to pqos as its mon-interval argument which defines monitoring sampling interval for RDT metrics in 100ms units, where 1=100ms. More info about pqos: https://github.com/intel/intel-cmt-cat/tree/master/pqos

After providing by pqos tool any output data, the plugin parses it immediately and adds to telegraf accumulator. Also, the plugin checks at every interval if desired processes change their PIDs association.